colm/transpile/transpile.lm

```diff
@@ -0,0 +1,265 @@
+namespace flux
+	include '../flux.lm'
+end
+
+include '../influxql.lm'
+
+if argv->length < 1 {
+	send stderr "usage: ./transpile bucket-name
+	exit(1)
+}
+
+parse P: influxql [stdin] 
+if ( !P ) {
+	send stderr "error parsing influxql query: [error]
+	exit( 1 )
+}
+
+bool time_constraint( Any: any )
+{
+	for UE: unary_expr in Any {
+		if match UE "time"
+			return true
+	}
+	return false
+}
+
+def time_range
+	Start: flux::expression
+	Stop: flux::expression
+[]
+
+void unquote_times( AE: ref<additive_expression> )
+{
+	for QI: quoted_identifier in AE {
+		Time: str = $QI
+		Time = prefix( suffix( Time, 1 ), Time.length - 2 )
+		QI.data = Time
+	}
+}
+
+bool check_time( Range: ref<time_range>, CE: comparison_expression )
+{
+	if match CE "time [Op: comparison_operator Rhs: additive_expression ]"
+			&& ( match Op "<" || match Op "<=" )
+	{
+		unquote_times( Rhs )
+		Range.Stop = parse flux::expression [Rhs]
+		return true
+	}
+
+	if match CE "time [Op: comparison_operator Rhs: additive_expression ]"
+			&& ( match Op ">" || match Op ">=" )
+	{
+		unquote_times( Rhs )
+		Range.Start = parse flux::expression [Rhs]
+		return true
+	}
+
+	return false
+}
+
+# Find time constraints and verify we can handle them.
+void yank_time( Range: ref<time_range>, Expr: ref<expr> )
+{
+	Error: bool = false
+	for CE: comparison_expression in Expr {
+		if ( time_constraint( CE ) ) {
+			if !match CE "( [expr] )" {
+				if !check_time( Range, CE ) {
+					send stderr "ERROR: wonky time constraint: [CE]
+					Error = true
+				}
+
+				CE = cons comparison_expression " __ERASE_ME__ "
```

Comment:
What does this do?

Reply:
It's parsing "__ERASE_ME__" as a comparison expression (at compile time) and then replacing the comparison expression it visited in the surrounding for loop with the resulting tree.

The thing to remember: `cons` is a compile time parse. It must succeed at compile time and all constituents must be valid types. You cannot supply it text at runtime.

`parse` is a run-time parse. It takes text as input. It can fail at run-time.